packages/xstate-solid/src/fsm.ts

```diff
@@ -0,0 +1,106 @@
+  TState extends Typestate<TContext> = { value: any; context: TContext }
+>(
+  service: StateMachine.Service<TContext, TEvent, TState>
+): StateMachine.State<TContext, TEvent, TState> => {
+  let currentValue: StateMachine.State<TContext, TEvent, TState>;
+  service
+    .subscribe((state) => {
+      currentValue = state;
+    })
+    .unsubscribe();
+  return currentValue!;
+};
+
+export function useMachine<TMachine extends StateMachine.AnyMachine>(
+  machine: TMachine,
+  options?: MachineImplementationsFrom<TMachine>
+): [StateFrom<TMachine>, ServiceFrom<TMachine>['send'], ServiceFrom<TMachine>] {
+  const resolvedMachine = createMachine(
+    machine.config,
+    options ? options : (machine as any)._options
+  );
+
+  const service = interpret(resolvedMachine).start();
+  const send = (event: ServiceFrom<TMachine>['send']) => service.send(event);
+
+  const [state, setState] = createStore<StateFrom<TMachine>>({
+    ...service.state,
+    matches(...args: Parameters<StateFrom<TMachine>['matches']>) {
+      // tslint:disable-next-line:no-unused-expression
+      (state as StateFrom<StateMachine.AnyMachine>).value; // sets state.value to be tracked by the store
+      return service.state.matches(args[0] as never);
+    }
+  } as StateFrom<TMachine>);
+
+  onMount(() => {
+    service.subscribe((nextState) => {
+      batch(() => {
+        updateState(nextState as StateFrom<TMachine>, setState);
+      });
+    });
+
+    onCleanup(service.stop);
+  });
+
+  return [state, send, service] as any;
+}
+
+export function useService<TService extends StateMachine.AnyService>(
+  service: Accessor<TService>
+): [StateFrom<TService>, TService['send'], TService] {
+  const serviceState = getServiceState(service());
+  const [state, setState] = createStore(serviceState);
+
+  const send = (event: TService['send']) => service().send(event);
+
+  createEffect(
+    on(service, (_, prev) => {
+      if (prev) {
+        checkReusedService(getServiceState(service()));
+      }
+      const { unsubscribe } = service().subscribe((nextState) => {
+        setState(
+          reconcile<typeof nextState, typeof nextState>(nextState, {
+            merge: false
+          })
+        );
+      });
+      onCleanup(unsubscribe);
+    })
+  );
+
+  return [state, send, service()] as any;
+}
+
+function checkReusedService(serviceState: StateMachine.State<any, any, any>) {
+  // Check if the machine has already been used as a service
+  // When FSM services re-use a single machine they share a reference to the same
+  // machine state object, causing updates to propagate to all services when used
+  // with Solid's store
+  if ($PROXY in serviceState) {
+    // tslint:disable-next-line:no-console
+    console.warn(
+      'Reusing an FSM machine will cause unexpected state changes in @xstate/solid.\n' +
+        'Use a factory function to reuse a machine: `const newMachine = () => createMachine(...some machine config)`'
+    );
+  }
```

Comment:
IIRC this is new - in what situation this warning is being raised? How the user's code would look like for this to be triggered?

Reply:
The situation that warrants this isn't new, I detailed it in [this comment above](https://github.com/statelyai/xstate/pull/2932#discussion_r888594979)

> For FSM services, in `@xstate/fsm` a reused machine has the [same State object reference](https://codesandbox.io/s/xstate-fsm-same-reference-dlhsev?file=/src/index.js). So any changes to a machine will propagate to all instances that are using that machine.
> 
> Here's a very simple example to show deep tracking in the store: https://codesandbox.io/s/solid-deep-store-mel80s?file=/src/index.tsx
>

Reply:
Here is a codesanbox showing the problem: https://codesandbox.io/s/wonderful-flower-sd9lmh?file=/src/index.tsx

If you hit the "INC" button a few times then switch to "Service 2", the context from "Service 1" matches in "Service 2"

Reply:
Ok, I've debugged this a little bit and I see what happens. It's quite weird because something that is really basic doesn't work "as expected":
```js
const initial = { count: 0 }

const [state1, setState1] = createStore(initial)
const [state2, setState2] = createStore(initial)
```

This leads to the same problem as those states are not isolated from each other.

I wonder though - won't we have the same problem with the integration targeting the main XState library? Even though the `.state` might be of a different identity, some parts of it might share references. Imagine the simplest:
```ts
const m = createMachine({
  context: {
    count: 0
  }
})

const s1 = interpret(m).start()
const s2 = interpret(m).start()

s1.state === s2.state // false
s1.context === s2.context // true
```

Reply:
Yeah, this is why I ultimately resorted to the [rfdc](https://www.npmjs.com/package/rfdc) library to clone the state object. The majority of cases won't be an issue, even if two machines have a share a reference to an object, updates through assign will create a new object reference. Still, there are edge cases where this might pop up. 

I've added a few tests for the core hooks, they all pass when machines have a shared object reference. It's only FSM that fails these situations.

Reply:
Hm, OK - I don't quite understand this difference then 😅 I trust the tests but I'm really curious about the inner workings of this thing.